packages/devtools_app/web/flutter_bootstrap.js

```diff
@@ -21,14 +21,66 @@ function unregisterDevToolsServiceWorker() {
   }
 }
 
+// Calls the DevTools server API to read the user's wasm preference.
+async function getDevToolsWasmPreference() {
+  const request = 'api/getPreferenceValue?key=experiment.wasm';
+  try {
+    const response = await fetch(request);
+    if (!response.ok) {
+      console.warn(`[${response.status} response] ${request}`);
+      return false;
+    }
+
+    // The response text should be an encoded boolean value ("true" or "false").
+    const wasmEnabled = JSON.parse(await response.text());
+    return wasmEnabled === true || wasmEnabled === 'true';
+  } catch (error) {
+    console.error('Error fetching experiment.wasm preference value:', error);
+    return false;
+  }
+}
+
+// Gets the renderer ('canvaskit' or 'skwasm') for the Flutter bootstrap config
+// based on the value of the 'wasm' query parameter or the wasm setting from the
+// DevTools preference file. This method also updates the 'wasm' query parameter
+// to reflect the renderer that will be used.
+async function getRendererAndUpdateQueryParameter() {
+  // This query parameter must match the String value specified by
+  // `DevToolsQueryParameters.wasmKey`. See
+  // devtools/packages/devtools_app/lib/src/shared/query_parameters.dart
+  const wasmQueryParameterKey = 'wasm';
+
+  const searchParams = new URLSearchParams(window.location.search);
+  const wasmEnabledFromQueryParameter = searchParams.get(wasmQueryParameterKey) === 'true';
+  const wasmEnabledFromDevToolsPreference = await getDevToolsWasmPreference();
+  const shouldUseSkwasm = wasmEnabledFromQueryParameter === true || wasmEnabledFromDevToolsPreference === true;
+  
+  const url = new URL(window.location.href);
+  // Ensure the 'wasm' query parameter in the URL is accurate for the renderer
+  // DevTools will be loaded with.
+  if (shouldUseSkwasm) {
+    url.searchParams.set(wasmQueryParameterKey, 'true');
+  } else {
+    url.searchParams.delete(wasmQueryParameterKey);
+  }
+  // Update the browser's history without reloading. This is a no-op if the wasm
+  // query parameter does not actually need to be updated.
+  window.history.pushState({}, '', url);
+
+  return shouldUseSkwasm ? 'skwasm' : 'canvaskit';
```

Comment:
One note: if you actually specify `skwasm` and it fails to load for some reason, you won't get JS fallback (because we will be unable to fulfill your request for the skwasm renderer). So if the behavior you want is to use the skwasm renderer if possible but otherwise fall back to JS, just don't specify the renderer at all (it's the default behavior).

Reply:
Done.